Added an internal solution to problem "GCD on Blackboard" in the silver "Introduction to Prefix Sums" section. (#1886)
cpinitiative/usaco-guide: solutions/ac-GCDOnBlackboard.mdx

---
id: ac-GCDOnBlackboard
source: AC
title: GCD On Blackboard
author: Andrew Cheng
---

**Time Complexity**: $\mathcal{O}(N \log N)$ (GCD operations take $\mathcal{O}(\log N)$ time)

> **Main Idea:**

Reviewer: ```suggestion
## Solution
```

> The question in its essence asks us to find the maximum possible GCD of the remaining $N-1$ numbers

Reviewer: ```suggestion
The question asks us to find the maximum possible GCD of the remaining $N-1$ numbers
```

> after taking any one of them away. Though naively trying every single combination will result in a

Reviewer: ```suggestion
after taking any one of them away. Naively trying every single combination will result in a
```

complexity of $\mathcal{O}(N^2)$ as there are $N$ different ways to take away a number and calculating
the GCD will take $\mathcal{O}(N)$ time per case. 

> To speed this process up, we can utilize a prefix sum and a suffix sum. Let $l[i]$ = $GCD_{j=1}^{i} a[j]$ and $r[i]$ = $GCD_{j=i}^{N} a[j]$.

Reviewer 2: Use `$\gcd$` instead of `$GCD$`

Then the answer is the maximum of $GCD(l[i-1],r[i+1]),$ $i \in [2,N-1]$. Though remember to check for the case

> where the first or the last number is replaced. (line 25 of code)

Reviewer: ```suggestion
where the first or the last number is replaced (line 25).
```
replace this with the right number after the below reviews

> **Solution:**

Reviewer: ```suggestion
## Implementation
```

<LanguageSection>
<CPPSection>

```cpp

#include <bits/stdc++.h>
using namespace std;

> const int maxN = 1e5+5;

Reviewer 3: wait wouldn't 1e5 + 1 work just sayin

Author: Uhhh I usually just leave myself some extra space. But no problem I can change it and the code should work just fine.

Reviewer 4: I personally use 1e5+10.... It's fine as is -- did not need to change this.

int a[maxN];
int l[maxN]; //l[i] = GCD of a1,a2, ..., ai
int r[maxN]; //r[i] = GCD of ai, ai+1, ..., an

> int N;

Reviewer 3: more intuitive variable names
i'd suggest `arr`, `prefGcd`, `suffGcd`, and `size` respectively
this way the names are self-documenting

Author: Using "size" instead of "N" will actually cause CE on the AtCoder server as "size" results in a conflict with a random system variable. I will opt for "N" instead of "size" here. The other suggestions are all adapted.

int gcd(int a, int b){

> 	if(b == 0)return a;
> 	return gcd(b,a%b);

Reviewer 3: ```suggestion
	return b == 0 ? a : gcd(b, a % b);
```
another option would be to just, you know, use `std::gcd`

Author: I wrote this in C++11. I think std::gcd is a thing since C++17 ?

Reviewer 3: wait does AC not support c++17 or something
i think most OJs do now, no?

Author: I'm not exactly sure if AC supports C++17 or not. I used C++11 cuz when I started USACO USACO only supported C++11.

Reviewer 4: It's fine to have it as 2 lines if you want.

}

> signed main(){

Reviewer 3: ```suggestion
int main() {

	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cin >> N;
	for(int i = 1; i <= N; ++i)cin >> a[i];
	l[1] = a[1]; r[N] = a[N];
	for(int i = 2; i <= N; ++i)l[i] = gcd(l[i-1],a[i]);
	for(int i = N-1; i >= 1; --i)r[i] = gcd(r[i+1],a[i]);

> 	int res = 0;

Reviewer 3: pplease add some whitespace separating relevant blocks of code, this physically hurts to look at

Reviewer 4: I think it's fine as is

	res = max(l[N-1],r[2]);
	for(int i = 2; i < N; ++i)res = max(res, gcd(l[i-1],r[i+1]));
	cout << res << '\n';

> 	return 0;

Reviewer 3: ```suggestion
```

Reviewer 4: Why?

---
id: ac-GCDOnBlackboard
source: AC
title: GCD On Blackboard
author: Andrew Cheng
---

**Time Complexity:**

> $\mathcal{O}(N+\log(max \ a_i))$

Reviewer 2: `\max` everywhere

Reviewer 2: can you explain why this is the time complexity :)

Author: My understanding is that the GCD operation reaches the maximum time complexity of ```O(log \max a_i)``` when GCD(a,b) = 1 (consecutive numbers in the fibonacci sequence if I recall correctly). Throughout each case in the brute force algorithm, the lowest possible GCD of the remaining ```N-1``` numbers is 1. Since it takes ```O(log \max a_i)``` operations to get the GCD to 1, the total time complexity of each case using brute force should be ```O(N + log(\max a_i)``` instead of ```O(Nlog(\max a_i))```. Is this correct ?

Reviewer 2: basically yes, this explanation is more complete though: https://codeforces.com/blog/entry/92720?#comment-814416

---
id: ac-GCDOnBlackboard
source: AC
title: GCD On Blackboard
author: Andrew Cheng
---

**Time Complexity:**
$\mathcal{O}(N \log N)$ (GCD operations take $\mathcal{O}(\log N)$ time)

## Solution
The problem asks us to find the maximum possible GCD of the remaining $N-1$ numbers
after taking any one of them away. Naively trying every single combination will result in a
complexity of $\mathcal{O}(N^2)$ as there are $N$ different ways to take away a number and calculating

> the GCD will take $\mathcal{O}(N)$ time per case. 

Reviewer 2: Why do you have O(N) here but O(N log N) above?

Actually, shouldn't the number of modulo operations be `O(N+\log(max a_i))`?

Author: My bad. Though shouldn't the time per case be ```O(N \log(max a_i))``` instead of ```O(N+\log(max a_i)) using brute force.

Reviewer 2: There are only `log(max a_i)` distinct prefix GCDs though

Author: Uhh I don't quite understand why are there only ```log(max a_i)``` distinct prefix GCDs. I was thinking that the basic method would be to try out all N different numbers to take away and for each case one would have to calculate the GCD of the current case by scanning across the rest N-1 numbers for a total complexity of ```O(N \log(max a_i))``` per case.

Reviewer 2: figure it out :)

> To speed this process up, we can utilize a prefix sum and a suffix sum. Let $l[i] = \gcd_{j=1}^{i} a[j]$ and $r[i]$ = $\gcd_{j=i}^{N} a[j]$.

Reviewer 2: ```suggestion
To speed this process up, we can the GCDs of every prefix and suffix. Let $l[i] = \gcd_{j=1}^{i} a[j]$ and $r[i]$ = $\gcd_{j=i}^{N} a[j]$.
```

Reviewer 2: Also see nwatx's earlier comment regarding the `=` sign.

> Then the answer is the maximum of $\gcd(l[i-1],r[i+1]),$ $i \in [2,N-1]$. 

Reviewer 2: ```suggestion
Then the answer is the maximum of $\gcd(l[i-1],r[i+1]),$ $i \in [1,N]$. 
```

## Implementation

<LanguageSection>
<CPPSection>

```cpp

#include <bits/stdc++.h>
using namespace std;

const int maxN = 1e5 + 5;

int arr[maxN];
int prefGcd[maxN]; //prefGcd[i] = GCD of a1,a2, ..., ai
int suffGcd[maxN]; //suffGcd[i] = GCD of ai, ai+1, ..., an
int N;

> int gcd(int a, int b){

Reviewer 2: Since AC and USACO support C++17, don't see a reason not to use `std::gcd`.

> 	return b == 0 ? a : gcd(b,a % b);

Reviewer 5: ```suggestion
	return b == 0 ? a : gcd(b, a % b);
```

}

int main(){
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cin >> N;
	for(int i = 1; i <= N; ++i)cin >> arr[i];

	prefGcd[0] = 0;
	suffGcd[N + 1] = 0;

	for(int i = 1; i <= N; ++i){

> 		prefGcd[i] = gcd(prefGcd[i - 1],arr[i]);
> 	}
> 
> 	for(int i = N; i >= 1; --i){
> 		suffGcd[i] = gcd(suffGcd[i + 1],arr[i]);

Reviewer 5: ```suggestion
		prefGcd[i] = gcd(prefGcd[i - 1], arr[i]);
	}

	for(int i = N; i >= 1; --i){
		suffGcd[i] = gcd(suffGcd[i + 1], arr[i]);
```

	}

	int res = 0;
	for(int i = 1; i <= N; ++i){

> 		res = max(res, gcd(prefGcd[i - 1],suffGcd[i + 1]));

Reviewer 5: ```suggestion
		res = max(res, gcd(prefGcd[i - 1], suffGcd[i + 1]));
```

---
id: ac-GCDOnBlackboard
source: AC
title: GCD On Blackboard
author: Andrew Cheng
---

## Time Complexity

> $\mathcal{O}(N \log N)$ (GCD operations take $\mathcal{O}(\log N)$ time)

Reviewer: see: https://usaco.guide/general/adding-solution?lang=cpp#steps on ordering of time complexity (and formatting)

## Solution

> The question asks us to find the maximum possible GCD of the remaining $N-1$ numbers

Reviewer 4: "question" ?? Probably should use task/problem

after taking any one of them away. Naively trying every single combination will result in a
complexity of $\mathcal{O}(N^2)$ as there are $N$ different ways to take away a number and calculating
the GCD will take $\mathcal{O}(N)$ time per case. 

> To speed this process up, we can utilize a prefix sum and a suffix sum. Let $l[i]$ = $\gcd_{j=1}^{i} a[j]$ and $r[i]$ = $\gcd_{j=i}^{N} a[j]$.

Reviewer: ```suggestion
To speed this process up, we can utilize a prefix sum and a suffix sum. Let $l[i] = \gcd_{j=1}^{i} a[j]$ and $r[i] = \gcd_{j=i}^{N} a[j]$.
```

Then the answer is the maximum of $\gcd(l[i-1],r[i+1]),$ $i \in [2,N-1]$. Though remember to check for the case

> where the first or the last number is replaced (line 25).

Reviewer: maybe add a comment instead of referring to line 25 (since suggestions might change its position)

> where the first or the last number is replaced (line 25).

Reviewer 4: Also, gcd(0, x) is always x, so replacing the first and last numbers can work w/o special cases by letting pref[0] = 0 and suff[N+1] = 0

## Implementation

<LanguageSection>
<CPPSection>

```cpp

#include <bits/stdc++.h>
using namespace std;

> const int maxN = 1e5+1;

Reviewer 5: ```suggestion
const int maxN = 1e5 + 1;
```
can you add spaces around all the other operators too?

---
id: ac-GCDOnBlackboard
source: AC
title: GCD On Blackboard
author: Andrew Cheng
---

> **Time Complexity:**

Reviewer 2: probably better to move all the time complexity stuff after line 34

Reviewer 6: ^

$\mathcal{O}(N+\log(\max a_i))$

<Spoiler title="Time Complexity Proof">

At first glance, one might think the task of finding the GCD of an array of integers will take $\mathcal{O}(N\log(\max a_i))$ time

> as each GCD operation takes atmost $\mathcal{O}(\log(\max a_i))$ time. However, as each GCD operation will decrease the maximum of $(a,b)$

Reviewer 2: whare are $a$ and $b$?

> as each GCD operation takes atmost $\mathcal{O}(\log(\max a_i))$ time. However, as each GCD operation will decrease the maximum of the two operands

Reviewer 2: ```suggestion
as each GCD operation takes at most $\mathcal{O}(\log(\max a_i))$ time. However, as each GCD operation will decrease the maximum of the two operands
```

by at least a factor of 2, there could only be a total of $\mathcal{O}(\log (\max a_i))$ GCD operations
as opposed to $\mathcal{O}(N \log(\max a_i))$ operations (The smallest possible GCD is 1 by definition). Hence the total time complexity of this problem is $\mathcal{O}(N+\log(\max a_i))$
as stated previously.

<Optional>

> This codeforces blog provides a more detailed explanation to taking the GCD of an array: https://codeforces.com/blog/entry/92720?#comment-814416

Reviewer 2: ```suggestion
[This](https://codeforces.com/blog/entry/92720?#comment-814416) CodeForces blog discusses the time complexity of finding the GCD of an array.
```

At first glance, one might think the task of finding the GCD of an array of integers will take $\mathcal{O}(N\log(\max a_i))$ time
as each GCD operation takes at most $\mathcal{O}(\log(\max a_i))$ time. However, as each GCD operation will decrease the maximum of the two operands
by at least a factor of 2, there could only be a total of $\mathcal{O}(\log (\max a_i))$ GCD operations
as opposed to $\mathcal{O}(N \log(\max a_i))$ operations (The smallest possible GCD is 1 by definition). Hence the total time complexity of this problem is $\mathcal{O}(N+\log(\max a_i))$
as stated previously.

<Optional>

[This](https://codeforces.com/blog/entry/92720?#comment-814416) CodeForces blog discusses the time complexity of finding the GCD of an array.

</Optional>

</Spoiler>

## Implementation

<LanguageSection>
<CPPSection>

```cpp

#include <bits/stdc++.h>
using namespace std;

const int maxN = 1e5 + 5;

int arr[maxN];
int prefGcd[maxN]; //prefGcd[i] = GCD of a1,a2, ..., ai
int suffGcd[maxN]; //suffGcd[i] = GCD of ai, ai+1, ..., an
int N;

> int main(){

Reviewer: ```suggestion
int main() {
```

	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cin >> N;

> 	for(int i = 1; i <= N; ++i)cin >> arr[i];

Reviewer: ```suggestion
	for(int i = 1; i <= N; ++i) cin >> arr[i];
```
